### PR TITLE
Remove implicit broadcasting from `Pauli`/label construction of `PauliList`

### PR DESCRIPTION
**Pull request by garrison**
### Summary

When constructing a `PauliList` from `Pauli`s or labels, the number of qubits is inferred from the first argument, and a `ValueError` is raised if any subsequent element has a different number of qubits.  For instance:

```python
>>> from qiskit.quantum_info import PauliList
>>> PauliList(["XXX", "ZZ"])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/garrison/Qiskit/qiskit-terra3/qiskit/quantum_info/operators/symplectic/pauli_list.py", line 140, in __init__
    base_z, base_x, base_phase = self._from_paulis(data)
                                 ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/garrison/Qiskit/qiskit-terra3/qiskit/quantum_info/operators/symplectic/pauli_list.py", line 193, in _from_paulis
    base_z[i] = pauli._z
    ~~~~~~^^^
ValueError: could not broadcast input array from shape (2,) into shape (3,)
```

However, if any _single_-qubit `Pauli` or label is provided, it will be implicitly broadcast to the correct number of qubits:

```python
>>> PauliList(["XXX", "Z", "I", "ZYX"])
PauliList(['XXX', 'ZZZ', 'III', 'ZYX'])
```

This PR removes this implicit broadcasting behavior and makes this expression raise a `ValueError` in all cases of mismatched qubit count.  It also provided a clearer error message in all such cases.

### Details and comments

I originally "snuck" this change into #9770, but I realized it is better to pull it out into a separate PR, as it deserves its own discussion (and release note).

IMO this should be considered a bug fix, as the broadcasting behavior is not documented.  (I only accidentally stumbled upon it after writing a test for #9770 which unexpectedly did not raise a `ValueError`.)  In the unlikely case that somebody, somewhere, is relying on this behavior, I think seeing an informative error message upon upgrading to 0.24 will be a fine outcome, i.e., I don't believe a deprecation cycle is necessary.  However, an argument _could_ be made in favor of it.  Either way, I would hesitate to include this change as part of a stable backport.

### CODE REVIEW

File: releasenotes/notes/paulilist-do-not-broadcast-from-paulis-96de3832fba21b94.yaml
Comment (jakelishman) [inline releasenotes/notes/paulilist-do-not-broadcast-from-paulis-96de3832fba21b94.yaml:5]:
```suggestion
    :class:`.PauliList` is initialized from :class:`.Pauli`\ s or labels.  For
```

(just to get the Sphinx crossrefs, and docutils needs funny things when using letters after backticks)

**Review by jakelishman — COMMENTED** [review]
This looks sensible to me: it definitely feels like `PauliList(["XXX", "Z"])` should be flagged as user error.  This should wait for @ikkoham's approval too.

**Review by jakelishman — APPROVED** [review]
Looks good to me, but I'll hold off merge to wait for Ikko.

**Comment by 1ucian0** [comment]
ping @ikkoham

**Review by ikkoham — APPROVED** [review]
LGTM! Thank you!

**Comment by jakelishman** [comment]
@Mergify refresh

**Comment by jakelishman** [comment]
@Mergify requeue

**Comment by jakelishman** [comment]
Manually updating because the mergify bot seems to have got itself stuck somehow.

**Pull request merged**
